test/client-spec.js

```diff
@@ -103,6 +110,44 @@ describe('Client', function() {
         client.connect();
     });
 
+    it('should not throw uncaught exception on initial websocket connection timeout', function(done) {
+
+        var client = new Client({
+            url: brokenMockServerUrl,
+            reconnect: true,
+            connectionTimeout: 200,
+            handshakeTimeout: 300
+        });
+        var handshakeTimeoutError = 'Opening handshake has timed out';
+        var earlyCloseError = 'WebSocket was closed before the connection was established';
+        var caughtError = false
+        var isDone = false
+
+        function onUncaughtException(err) {
+            if (isDone) { return; }
+            isDone = true;
+            process.off('uncaughtException', onUncaughtException);
+            var msg = err.message;
+            done(new Error('Client should not cause uncaught exception: ' + msg));
+        }
+
+        client.onError((err) => {
+            if (isDone) { return; }
+            var msg = err.message;
+            caughtError = msg === handshakeTimeoutError || msg === earlyCloseError;
+            if (caughtError) {
+                isDone = true;
+                process.off('uncaughtException', onUncaughtException);
+                done();
+            }
+
+        });
+
+        process.on('uncaughtException', onUncaughtException)
+
+        client.connect();
+    });
```

Comment:
I'm going to split this test into two tests that cover when the handshakeTimeout occurs before and after the connectionTimeout.